File: src/main/java/fds/java/events/ConsumerWriter.java
package fds.java.events;


import com.zink.queue.Connection;
import com.zink.queue.ConnectionFactory;
import com.zink.queue.ReadChannel;
import io.nats.streaming.StreamingConnection;
import io.nats.streaming.StreamingConnectionFactory;


/**
 * Stub for the Consumer of information from the Queue and writer into the event store
 *
 * The following condensed from Docker
 *
 * To Pull Nats server from Docker
 * > docker pull nats-streaming
 *
 * Then to run on Windows
 * > docker run -d -p 4222:4222 -p 8222:8222 nats-streaming nats-streaming-server -p 4222 -m 8222
 *
 *  And on Unix
 * > docker run -d -p 4222:4222 -p 8222:8222 nats-streaming -p 4222 -m 8222
 *
 * To check it is running
 * > docker ps
 *
 */
public class ConsumerWriter {


    public static void main(String[] args) throws Exception {

        // Connect to queue
        final String ipAddr = "192.168.1.84";
        final Connection con = ConnectionFactory.connect(ipAddr);
        final String channelName = "BBC7";
        final ReadChannel rc = con.subscribe(channelName);

        // Connect to EventStore
        final String clusterID = "test-cluster";
        final String clientID = "event-writer";
        final StreamingConnectionFactory cf = new StreamingConnectionFactory(clusterID, clientID);
        final StreamingConnection sc = cf.createConnection();

        // TODO - write the events into the event store

        // Hints ...

        // to read an item from the queue
        final String event = (String) rc.read();

        // to write an item to the event store
        sc.publish(channelName, event.getBytes());
	}


}

File: src/main/scala/fds/scala/events/ConsumerWriter.scala
package fds.scala.events


import io.nats.client.{Connection, Dispatcher, Nats}
import io.nats.streaming.{Message, StreamingConnection, StreamingConnectionFactory}

import java.nio.charset.StandardCharsets


/**
  * Stub for the Consumer of information from the Queue and writer into the EventStore
  */
object ConsumerWriter {

  @throws[Exception]
  def main(args: Array[String]): Unit = {

    // Connect to the message queue
    val nc: Connection = Nats.connect("nats://localhost:4222")

    // Connect to EventStore
    val clusterID = "test-cluster"
    val clientID = "event-writer"
    val cf: StreamingConnectionFactory = new StreamingConnectionFactory(clusterID, clientID)
    val sc: StreamingConnection = cf.createConnection

    val dispatcher = nc.createDispatcher((msg) => {
                // Check message arrived
                println(new String(msg.getData(), StandardCharsets.UTF_8))
                // use sc to publish the message
                sc.publish("log-events", msg.getData())
          });
    dispatcher.subscribe("logs")
  }

}
